### ISSUE TITLE
DHCP client does not send the correct hostname

### ISSUE THREAD

**Issue by jhberthemet** [boot2docker/boot2docker#1281]
I'm using boot2docker with docker-machine create on ESXi. When creating a VM using DHCP everything works well however udhcpc registers itself as boot2docker hostname instead of the real hostname.

I see that dhcp.sh is correctly sending the request:
`    /sbin/udhcpc -b -i $DEVICE -x hostname:$(/bin/hostname) -p /var/run/udhcpc.$DEVICE.pid >/dev/null 2>&1 &
 `

So I guess it's invoked before the hostname is set. Could you give me advice on how to change that in boot2docker iso?

**Comment by jhberthemet** [comment]
I found /opt/bootscript.sh and hostname command is called before dhcp.sh, so the order is correct but result is not.

**Comment by jhberthemet** [comment]
I managed to connect before hostname was executed and hostname was 'box' whereas it was boot2docker in DNS, so hostname was changed but I don't get where this name comes from.

**Comment by tarekch** [comment]
What i noticed is that it tries first with boot2docker or box and then after it figures out its username, it is not trying again. probably cuz it assumes that it has an IP already so why bother. Either check the dhcp.sh file to see if there is something stopping it or do what i did, add this line to the end of ur bootscript.sh file: `/sbin/udhcpc -F $(/bin/cat /etc/hostname) -i eth1`

**Comment by jhberthemet** [comment]
Thanks for the suggestion, I'll try that when I'll have some time.

It would be great to have this change made by default in boot2docker.

**Comment by tianon** [comment]
Is this still an issue on the latest release? (which is using the latest version of the DHCP code from TCL 10.x upstream)

**Comment by KyleRogers** [comment]
Yes, the issue still persists. I'm using docker-machine to spin up a virutalbox vm. The vm registers itself as boot2docker on my DHCP which propagates to my internal DNS.

```
docker@production2:~$ uname -a 
Linux production2 4.14.141-boot2docker #1 SMP Wed Sep 4 19:02:58 UTC 2019 x86_64 GNU/Linux
```

**Comment by tianon** [comment]
I *really* hope you're not using boot2docker for actual production systems :sweat_smile: (https://github.com/boot2docker/boot2docker#caveat-emptor)

You're likely on the latest release (or at least plenty recent), I think this just means that whatever interaction we've got with TCL's DHCP script isn't working (or their script doesn't take hostname into account, but I find that much harder to believe).

Realistically, I think this is probably a timing issue -- I think their scripts do DHCP before our `bootsync.sh` runs to set the hostname:

https://github.com/boot2docker/boot2docker/blob/31e764375a64dd66a18ad3cab7aeeb35e659a532/files/bootsync.sh#L7-L13

So to make this work, we'll likely need to find (or create :scream: :disappointed: :sob:) an earlier entry point into the TCL boot process.

**Comment by rogermartensson** [comment]
What is the latest information regarding this issue?
I really would like to be able to register the correct hostname in my DNS when using DHCP.

**Comment by pombocaesar** [comment]
+1 on this one, is this issue even in the backlog?
